src/briefcase/commands/run.py

```diff
@@ -1,23 +1,217 @@
+import re
+import signal
 from abc import abstractmethod
 from typing import Optional
 
 from briefcase.config import BaseConfig
-from briefcase.exceptions import BriefcaseCommandError
+from briefcase.exceptions import BriefcaseCommandError, BriefcaseTestSuiteFailure
 
 from .base import BaseCommand, full_options
 
 
+class LogFilter:
+    DEFAULT_SUCCESS_REGEX = (
+        # Unittest
+        # The a{123} groups are to support Android logging differences.
+        # when https://github.com/chaquo/chaquopy/issues/746 is resolved,
+        # those groups can be deleted.
+        r"(^-{65,}\n(?P<a1> \n)?Ran \d+ tests in \d+\.\d{3}s\n(?P<a2> )?\n(?P<a3> \n)?OK( \(.*\))?$)"
+        # PyTest
+        r"|(^={10,} (((\d+ passed)?((, )?\d+ skipped)?(, \d+ warnings?)?)|(no tests ran)) in \d+\.\d+s ={10,}$)"
+    )
+
+    DEFAULT_FAILURE_REGEX = (
+        # Unittest
+        # The a{123} groups are to support Android logging differences.
+        # when https://github.com/chaquo/chaquopy/issues/746 is resolved,
+        # those groups can be deleted.
+        r"(^-{65,}\n(?P<a1> \n)?Ran \d+ tests in \d+\.\d{3}s\n(?P<a2> )?\n(?P<a3> \n)?FAILED( \(.*\))?$)"
+        # Pytest
+        r"|(^={10,} ((\d+ failed(, \d+ passed)?(, \d+ skipped)?)|(\d+ errors?)) in \d+.\d+s ={10,}$)"
+    )
+
+    def __init__(
+        self,
+        log_popen,
+        clean_filter=None,
+        clean_output=True,
+        success_filter=None,
+        failure_filter=None,
+    ):
+        """Create a filter for a log stream.
+
+        :param log_popen: The Popen object for the stream producing the logs.
+        :param clean_filter: (Optional) A function that will filter a line of
+            logs, returning a "clean" line without any log system preamble.
+        :param clean_output: Should the output displayed to the user be the
+            "clean" output? (Default: True).
+        :param success_filter: (Optional) A function that will operate on a
+            string containing the last 5 lines of "clean" (i.e., preamble
+            filtered) logs, returning True if a "success" condition has been
+            detected. If the success filter returns True, the log process will
+            be terminated.
+        :param success_filter: (Optional) A function that will operate on a
+            string containing the last 5 lines of "clean" (i.e., preamble
```

Comment:
I'm not sure a named constant helps here, because one of the uses is documentation. To make that code comment useful, we'd need to document the constant separately, or make the history size configurable; and both of those options seem like overkill for a class that is used entirely internally to this code module.

I've addressed the discrepancy; but my preference would be to leave it at that.